flink-table-store-connector/src/main/java/org/apache/flink/table/store/connector/source/FileStoreSource.java

```diff
@@ -122,6 +150,17 @@ public SplitEnumerator<FileStoreSourceSplit, PendingSplitsCheckpoint> restoreEnu
             SplitEnumeratorContext<FileStoreSourceSplit> context,
             PendingSplitsCheckpoint checkpoint) {
         FileStoreScan scan = fileStore.newScan();
+        Long snapshotId;
+        Collection<FileStoreSourceSplit> splits;
+        if (specifiedSnapshotId != null) {
+            Preconditions.checkNotNull(
+                    specifiedManifestEntries,
+                    "The manifest entries cannot be null for manual compaction.");
```

Comment:
The error message looks quite strange for me because I'm just curious :  Is it the only case that we will set `specifiedSnapshotId` and `specifiedManifestEntries` for manual compaction ?  Will be other case that we will set the two fields ?

Reply:
> The error message looks quite strange for me because I'm just curious : Is it the only case that we will set `specifiedSnapshotId` and `specifiedManifestEntries` for manual compaction ? Will be other case that we will set the two fields ?

So far, only manually invoked compaction (triggered by `ALTER TABLE ... COMPACT`) will specify the snapshot id and manifest entries. Other conditions will perform a scan during the runtime.

The reason is that `ALTER TABLE ... COMPACT` will pre-scan the latest snapshot during the planning phase, serialize the info as a string, and put it back to enriched options. Therefore, at the runtime, the source does not need to perform a scan anymore.

Reply:
Let me give a detailed explanation about 
> `ALTER TABLE ... COMPACT` will pre-scan the latest snapshot during the planning phase

`ALTER TABLE ... COMPACT` will be first converted to a `SinkModifyOperation` with `SourceQueryOperation` as the only child (on the Flink side). And then the `ManagedTableFactory#onCompactTable` is invoked, the impl(back to the TableStore) will perform a scan, collected manifest entries and the corresponding snapshot id, and serialized as a string. It will be put back to enriched options. So at the runtime, when the source is initialized, the splits can be directly generated from the options. This PR aims to pave the way for skipping runtime scans for this condition.